### PR TITLE
WT-9889 Reduce the overall number of cursor bound comparisons.

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

File: src/btree/bt_cursor.c [hunk 1]
```diff
@@ -422,12 +423,13 @@ __wt_cursor_valid(WT_CURSOR_BTREE *cbt, WT_ITEM *key, uint64_t recno, bool *vali
               session, cbt->ref->page, &cbt->ref->page->pg_row[cbt->slot], cbt->tmp, true));
             key = cbt->tmp;
         }
-
-        WT_RET(__btcur_bounds_contains_key(
-          session, &cbt->iface, key, WT_RECNO_OOB, &key_out_of_bounds, NULL));
-        /* The key value pair we were trying to return weren't within the given bounds. */
-        if (key_out_of_bounds)
-            return (0);
+        if (check_bounds) {
```
Comment (clarissecheah) [inline src/btree/bt_cursor.c:426]: This comment is addressing line 384. Would this call to `__btcur_bounds_contains_key `also require a `check_bounds` guard?

File: src/btree/bt_cursor.c [hunk 2]
```diff
@@ -2296,14 +2299,56 @@ __wt_btcur_bounds_position(
...
+            if (next)
+                /*
+                 * We need to check if the upper bound is being honored. We could optimize the call
+                 * to cursor valid and have it only check a single bound, but we would need to
+                 * indicate the bound and direction.
+                 */
+                WT_RET(__wt_cursor_valid(cbt, NULL, WT_RECNO_OOB, &valid, true));
+            else
+                /*
+                 * We know that this isn't a valid as we're doing a bound position call on the upper
```
Comment (clarissecheah) [inline src/btree/bt_cursor.c:2323]: "this isn't a valid" should be "this isn't valid"

File: src/btree/bt_cursor.c [hunk 3]
```diff
@@ -2296,14 +2299,56 @@ __wt_btcur_bounds_position(
...
+                 * doesn't exist as if it does we could in theory walk onto that. However if it did
+                 * exist we should've returned it in the cursor row search and thus have had a cbt
+                 * compare value of 0. Either way we check the assumption later with an assertion.
+                 */
+                *need_walkp = true;
+        } else {
+            /* We landed to the left of search key. */
+            if (next)
+                /*
+                 * We know that this isn't a valid as we're doing a bound position call on the lower
```
Comment (clarissecheah) [inline src/btree/bt_cursor.c:2336]: Ditto.

File: src/btree/bt_cursor.c (hunk 1 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c:426]: We'll still add the check_bounds guard as that way its consistent.

File: src/btree/bt_cursor.c (hunk 3 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c]: [6096b21](https://github.com/wiredtiger/wiredtiger/pull/8309/commits/6096b21b1f569d91c83d5af8d2fafce7effe577b)

File: src/btree/bt_cursor.c (hunk 2 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c]: [6096b21](https://github.com/wiredtiger/wiredtiger/pull/8309/commits/6096b21b1f569d91c83d5af8d2fafce7effe577b)

File: src/btree/bt_cursor.c (hunk 1 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c:428]: [6096b21](https://github.com/wiredtiger/wiredtiger/pull/8309/commits/6096b21b1f569d91c83d5af8d2fafce7effe577b)

**Review by agorrod — APPROVED** [review]
I'm not sure I remember the bigger picture well enough to know confidently that this is complete, but the changes all make sense to me.

The comments in this change are great - thanks!

All that means is: this deserves other reviewers as well.

File: src/btree/bt_curnext.c [hunk 4]
```diff
@@ -959,6 +959,17 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
...
             ret = __wt_cursor_key_order_check(session, cbt, true);
+
+        if (need_walk) {
+            /*
+             * The bounds positioning code relies on the assumption that if we had to walk then we
+             * can't possibly have walked to the lower bound. We check that assumption here by
+             * comparing the lower bound with our current key or recno.
+             */
+            ret = __wt_compare_bounds(
+              session, cursor, &cbt->iface.key, cbt->recno, false, &key_out_of_bounds);
```
Comment (jiechenbo) [inline src/btree/bt_curnext.c:970]: In the case that the lower bound inclusive flag is set, the __wt_compare_bounds function will mark the cursor positioned at the lower_bound as still inside the bounds. I think we may have to do a direct check on the lower bound key itself or turn off inclusive flag temporarily.

File: src/btree/bt_curprev.c [hunk 5]
```diff
@@ -894,6 +894,17 @@ __wt_btcur_prev(WT_CURSOR_BTREE *cbt, bool truncating)
...
             ret = __wt_cursor_key_order_check(session, cbt, false);
+
+        if (need_walk) {
+            /*
+             * The bounds positioning code relies on the assumption that if we had to walk then we
+             * can't possibly have walked to the upper bound. We check that assumption here by
+             * comparing the upper bound with our current key or recno.
+             */
+            ret = __wt_compare_bounds(
+              session, cursor, &cbt->iface.key, cbt->recno, true, &key_out_of_bounds);
```
Comment (jiechenbo) [inline src/btree/bt_curprev.c:905]: ditto.

File: src/btree/bt_curnext.c (hunk 4 above)
Comment (luke-pearson) [reply src/btree/bt_curnext.c:970]: That's a good point, ideally I want to reuse the compare bounds function. I wonder if we have a flag set unset macro.

File: src/btree/bt_cursor.c [hunk 6]
```diff
@@ -1166,11 +1170,10 @@ __wt_btcur_insert(WT_CURSOR_BTREE *cbt)
          * If not overwriting, fail if the key exists, else insert the key/value pair.
          */
         if (!F_ISSET(cursor, WT_CURSTD_OVERWRITE) && cbt->compare == 0) {
-            WT_ERR(__wt_cursor_valid(cbt, cbt->tmp, WT_RECNO_OOB, &valid));
+            WT_ERR(__wt_cursor_valid(cbt, cbt->tmp, WT_RECNO_OOB, &valid, false));
             if (valid)
                 goto duplicate;
         }
-
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/btree/bt_cursor.c]: b2f1be78d

File: src/btree/bt_curprev.c (hunk 5 above)
Comment (luke-pearson) [reply src/btree/bt_curprev.c:912]: b2f1be78d

File: src/btree/bt_curnext.c (hunk 4 above)
Comment (luke-pearson) [reply src/btree/bt_curnext.c:977]: b2f1be78d

File: src/btree/bt_cursor.c [hunk 7]
```diff
@@ -2296,14 +2302,56 @@ __wt_btcur_bounds_position(
...
+            /*
+             * If our compare value is 0 we found the exact key we were looking for, if the bound we
+             * are searching for is inclusive we can return it without walking.
+             */
+            if (F_ISSET(
+                  cursor, next ? WT_CURSTD_BOUND_LOWER_INCLUSIVE : WT_CURSTD_BOUND_UPPER_INCLUSIVE))
+                WT_RET(__wt_cursor_valid(cbt, cbt->tmp, WT_RECNO_OOB, &valid, false));
+            else
+                *need_walkp = true;
+        } else if (cbt->compare > 0) {
```
Comment (jiechenbo) [inline src/btree/bt_cursor.c:2315]: what do you think combining the if statements as (cbt->compare > 0 && next) || (cbt->compare < 0 && !next)? We can reduce down to one else case too, and have less need for comments.

**Review by jiechenbo — APPROVED** [review]
LGTM! I just have one question.

File: src/btree/bt_cursor.c (hunk 7 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c:2315]: Sounds good to me.

Comment (luke-pearson) [reply src/btree/bt_cursor.c]: [94a0b7c](https://github.com/wiredtiger/wiredtiger/pull/8309/commits/94a0b7c7710e831d2048fa14103fc83ecd152ffb)

**Pull request merged**